### PR TITLE
Fix bugs related to zero components in sequence

### PR DESCRIPTION
**Pull request by ioana-blue**
Don't allow creation of sequences with no components (+ test). Guard size calculation just in case.

### CODE REVIEW

**Comment by ioana-blue** [comment]
@rabbah PG exemption?

**Comment by ioana-blue** [comment]
Fixes #1760

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 1]
```diff
@@ -128,7 +128,7 @@ protected[core] case class BlackBoxExec(override val image: String, code: Option
 }
 
 protected[core] case class SequenceExec(components: Vector[FullyQualifiedEntityName]) extends Exec(Exec.SEQUENCE) {
-    override def size = components.map(c => c.size).reduce(_ + _)
+    override def size = if (components.size > 0) components.map(c => c.size).reduce(_ + _) else (0 B)
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: What about:

```
components.map(c => c.size).reduceOption(_ + _).getOrElse(0.B)
```

?

Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: Or do we even need size calculation on Sequences? Isn't the "size" of the list bound by its count anyway? Could we always return 0 byte?

Comment (ioana-blue) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: didn't know about reduceOption. gracias.

**Comment by ioana-blue** [comment]
@markusthoemmes fixed it. PG exemption?

**Comment by ioana-blue** [comment]
PG3 209

**Comment by ioana-blue** [comment]
this is ready to merge as far as I'm concerned @markusthoemmes the ball is in your court

**Comment by rabbah** [comment]
@ioana-blue if this is pg approved, please add corresponding label.

**Comment by ioana-blue** [comment]
@rabbah can't add label, I'm just a contributor :)

**Comment by ioana-blue** [comment]
I can't do any assignments on the right hand side of this window... to be more precise

File: core/controller/src/main/scala/whisk/core/controller/Actions.scala [hunk 2]
```diff
@@ -632,4 +635,5 @@ trait WhiskActionsApi
 
 private case class BlockingInvokeTimeout(activationId: ActivationId) extends TimeoutException
 private case class TooManyActionsInSequence() extends RuntimeException
+private case class NoComponentInSequence() extends RuntimeException
```
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: should these be illegal state exceptions instead of runtime exceptions?

Comment (ioana-blue) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: debatable. IllegalStateException makes me think that I made a request at a bad time (bad state on the server side). these exception look more similar to IllegalArgumentException - I'm performing a request with arguments that don't fit the bill/contract. 

I don't feel strongly about this, so I can change them. what's it gonna be? @rr @markusthoemmes

Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Actions.scala]: IllegalArgumentException is what I meant - thanks. That's appropriate. RuntimeExceptions don't fit these.

**Comment by ioana-blue** [comment]
all done. anything else?

File: tests/src/whisk/core/controller/test/SequenceApiTests.scala [hunk 3]
```diff
@@ -73,6 +73,19 @@ class SequenceApiTests
...
     }
 
+    it should "reject creation of sequence with no component specified" in {
+        implicit val tid = transid()
+        val seqName = EntityName(s"${aname()}_no_component")
+        // create exec sequence with no component
+        val content = WhiskActionPut(Some(Exec.sequence(Vector())))
+
+        // create an action sequence
+        Put(s"$collectionPath/${seqName.name}", content) ~> sealRoute(routes(creds)) ~> check {
```
Comment (rabbah) [inline tests/src/whisk/core/controller/test/SequenceApiTests.scala]: should also test on update (?overwrite=true).

File: common/scala/src/main/scala/whisk/core/entity/Exec.scala [hunk 4]
```diff
@@ -128,7 +128,7 @@ protected[core] case class BlackBoxExec(override val image: String, code: Option
 }
 
 protected[core] case class SequenceExec(components: Vector[FullyQualifiedEntityName]) extends Exec(Exec.SEQUENCE) {
-    override def size = components.map(c => c.size).reduce(_ + _)
+    override def size = components.map(c => c.size).reduceOption(_ + _) getOrElse (0 B)
```
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Exec.scala]: @markusthoemmes once we limit the length of the action names, i'm ok turning this into a no-op (return `0.B`).

**Comment by ioana-blue** [comment]
PG3 225

File: tests/src/whisk/core/controller/test/SequenceApiTests.scala (hunk 3 above)
Comment (ioana-blue) [inline tests/src/whisk/core/controller/test/SequenceApiTests.scala]: added test

**Comment by rabbah** [comment]
Merged as https://github.com/openwhisk/openwhisk/commit/5ace5077581aa75ad349ef33a2acb9fc1730ccfc.

**Pull request closed without merging**